csrc/scheduler/matmul_utils.cpp

```diff
@@ -294,6 +315,110 @@ std::string isMatmulFusionDefinitionSupported(
   return "";
 }
 
+// Assume that tens has a contiguous dimension, and that we will load rows of
+// that dimension, without merging with another dimension first. Then determine
+// the maximum vectorization that can be used.
+//
+// If the argument has no contiguous dimensions, then a vectorization width of 1
+// is returned.
+//
+// These rows can start at any multiple of the non-contiguous strides, so we
+// seek the largest power of 2 that divides all those other dimensions (capped
+// to 16) as well as the data pointer.
+int64_t maxRowVectorization(const at::Tensor& tens) {
+  const int64_t data_ptr_int =
+      static_cast<int64_t>(reinterpret_cast<std::uintptr_t>(tens.data_ptr()));
+  int64_t vec_size = scheduler_utils::maxVectorizationWidth(data_ptr_int);
+  vec_size = std::min(vec_size, 16l);
+  vec_size /= (int64_t)tens.element_size();
+  vec_size = std::max(vec_size, 1l);
+  if (vec_size == 1l) {
+    return vec_size;
+  }
+
+  // Find innermost dimension
+  bool contiguous = false;
+  for (int64_t i : c10::irange(tens.ndimension())) {
+    int64_t stride = tens.stride(i);
+    int64_t size = tens.size(i);
+    if (stride == 1) {
+      // Innermost dimension
+      contiguous = true;
```

Comment:
Discontiguous (in the Fusion) tensors are supported properly now as of the latest push.
 
The overlapping case is more interesting. For the case of multiple dimensions with stride 1 and corresponding sizes that are _not_ 1, I think the vectorization should always be 1 since there is definitely a row that's equal to another offset by 1.  However, if the outer size is also 1 in that case then we could possibly still vectorize. Since it is somewhat difficult to determine which row will be vectorized at this point in the code, I think we should play it safe and set vectorization to 1 if there are multiple stride=1 dims.